src/joystick/SDL_gamepad.c

```diff
@@ -688,6 +709,359 @@ static GamepadMapping_t *SDL_CreateMappingForAndroidGamepad(SDL_GUID guid)
+    case 1:
+        SDL_ADD_BUTTON_MAPPING("misc3", current_button++, mapping_string_len);
+        break;
+
+    case 2:
+        SDL_ADD_BUTTON_MAPPING("misc3", current_button++, mapping_string_len);
+        SDL_ADD_BUTTON_MAPPING("misc4", current_button++, mapping_string_len);
+        break;
+
+    case 3:
+        SDL_ADD_BUTTON_MAPPING("misc3", current_button++, mapping_string_len);
+        SDL_ADD_BUTTON_MAPPING("misc4", current_button++, mapping_string_len);
+        SDL_ADD_BUTTON_MAPPING("misc5", current_button++, mapping_string_len);
+        break;
+
+    case 4:
+        SDL_ADD_BUTTON_MAPPING("misc3", current_button++, mapping_string_len);
+        SDL_ADD_BUTTON_MAPPING("misc4", current_button++, mapping_string_len);
+        SDL_ADD_BUTTON_MAPPING("misc5", current_button++, mapping_string_len);
+        SDL_ADD_BUTTON_MAPPING("misc6", current_button++, mapping_string_len);
+        break;
+
+    default:
+        break;
+    }
+}
+
+/*
+* Helper function to decode SInput features information packed into version
+*/
+static bool SDL_CreateMappingStringForSInputGamepad(Uint16 vendor, Uint16 product, Uint8 sub_product, Uint16 version, Uint8 face_style, char* mapping_string, size_t mapping_string_len)
+{
+    SDL_SInputStyles_t decoded = { 0 };
+
+    switch (face_style) {
+    default:
+        SDL_strlcat(mapping_string, "face:abxy,", mapping_string_len);
+        break;
+    case 2:
+        SDL_strlcat(mapping_string, "face:axby,", mapping_string_len);
+        break;
+    case 3:
+        SDL_strlcat(mapping_string, "face:bayx,", mapping_string_len);
+        break;
+    case 4:
+        SDL_strlcat(mapping_string, "face:sony,", mapping_string_len);
+        break;
+    }
+
+    // For a sub-product value of 0, we interpret the
+    // mapping string dynamically based on the feature responses
+    if (sub_product == 0) {
+        // Decode styles for correct dynamic features
+        decoded.misc_style = (SInput_MiscStyleType)(version % SINPUT_MISCSTYLE_MAX);
+        version /= SINPUT_MISCSTYLE_MAX;
+
+        decoded.touch_style = (SInput_TouchStyleType)(version % SINPUT_TOUCHSTYLE_MAX);
+        version /= SINPUT_TOUCHSTYLE_MAX;
+
+        decoded.meta_style = (SInput_MetaStyleType)(version % SINPUT_METASTYLE_MAX);
+        version /= SINPUT_METASTYLE_MAX;
+
+        decoded.paddle_style = (SInput_PaddleStyleType)(version % SINPUT_PADDLESTYLE_MAX);
+        version /= SINPUT_PADDLESTYLE_MAX;
+
+        decoded.trigger_style = (SInput_TriggerStyleType)(version % SINPUT_TRIGGERSTYLE_MAX);
+        version /= SINPUT_TRIGGERSTYLE_MAX;
+
+        decoded.bumper_style = (SInput_BumperStyleType)(version % SINPUT_BUMPERSTYLE_MAX);
+        version /= SINPUT_BUMPERSTYLE_MAX;
+
+        decoded.analog_style = (SInput_AnalogStyleType)(version % SINPUT_ANALOGSTYLE_MAX);
+
+        SDL_SInputStylesMapExtraction(&decoded, mapping_string, mapping_string_len);
+        return true;
+    }
+
+    // For non-zero sub-product IDs
```

Comment:
Do you set non-zero sub-product IDs for any controller? Do you even need this block, other than adding a comment on how to handle it?

Reply:
In that case you should set a specific sub_product below instead of using the default, so if someone modifies your controller and uses e.g. sub_product 2, they won't get your old mapping.

I would lean towards just using the dynamic mapping if your controller works well with it, and eliminating this confusion.